lib/rules/prefer-destructuring.js

```diff
@@ -109,24 +136,71 @@ module.exports = {
                 return;
             }
 
-            if (checkArrays && isArrayIndexAccess(rightNode)) {
-                report(reportNode, "array");
+            const sourceCode = context.getSourceCode();
+            const previousNode = sourceCode.getTokenBefore(reportNode);
+
+            // Check to avoid issues with ASI
+            if (previousNode && previousNode.type !== "Keyword" && previousNode.value !== ";") {
                 return;
             }
 
-            if (checkObjects && enforceForRenamedProperties) {
-                report(reportNode, "object");
-                return;
+            let type;
+
+            if (checkArrays && isArrayIndexAccess(rightNode)) {
+                type = "array";
             }
 
             if (checkObjects) {
                 const property = rightNode.property;
+                const hasNameMatch = ((property.type === "Literal" && leftNode.name === property.value) ||
+                                      (property.type === "Identifier" && leftNode.name === property.name));
 
-                if ((property.type === "Literal" && leftNode.name === property.value) ||
-                    (property.type === "Identifier" && leftNode.name === property.name)) {
-                    report(reportNode, "object");
+                if (hasNameMatch || enforceForRenamedProperties) {
+                    type = "object";
                 }
             }
+
+            if (!type) {
+                return;
+            }
+
+            context.report({
+                node: reportNode,
+                message: `Use ${type} destructuring`,
+                fix: fixer => {
+
+                    // No-op if destructuring is type array
+                    if (type === "array") {
+                        return null;
+                    }
+
+                    const lastToken = sourceCode.getLastToken(reportNode);
+                    const nodeText = sourceCode.getText(reportNode);
+
+                    const removalState = findRightHandRemovalState(lastToken, sourceCode);
+                    const removalIndex = nodeText.length - removalState.lengthToRemove;
+
+                    let fixedText;
+
+                    fixedText = nodeText.slice(0, removalIndex);
+
+                    // Remove the left node to replace with destructured text
+                    const noLeftNodeText = fixedText.slice(leftNode.name.length, fixedText.length);
+
+                    if (!enforceForRenamedProperties) {
```

Comment:
Could you clarify why the `enforceForRenamedProperties` option needs to be checked here? I don't understand why the option would need to affect the behavior of an autofix.